tests/unit/replication/ReplicationTaskScheduler.spec.js

```diff
@@ -1,39 +1,181 @@
 const assert = require('assert');
+const uuid = require('uuid/v4');
 
-const ReplicationTaskScheduler = require(
-    '../../../extensions/replication/utils/ReplicationTaskScheduler');
-
-describe('replication task scheduler', () => {
-    it('should ensure serialization of updates to the same versioned object',
-    done => {
-        const taskScheduler = new ReplicationTaskScheduler(
-            (entry, done) => {
-                setTimeout(() => {
-                    // eslint-disable-next-line no-param-reassign
-                    entry.object.value = entry.setValueTo;
-                    done();
-                }, Math.random() * 100);
-            });
-        const objects = [];
-        let doneCount = 0;
+const ObjectQueueEntry = require('../../../lib/models/ObjectQueueEntry');
+const ActionQueueEntry = require('../../../lib/models/ActionQueueEntry');
+
+const ReplicationTaskScheduler =
+    require('../../../extensions/replication/utils/ReplicationTaskScheduler');
+
+function getMockQueueEntry(params) {
+    const { objectKey, versionId, contentMD5 } = params;
+    return new ObjectQueueEntry('test-bucket-name', objectKey)
+        .setVersionId(versionId)
+        .setContentMd5(contentMD5 || 'd41d8cd98f00b204e9800998ecf8427e');
+}
+
+function getMockActionEntry(params) {
+    const { objectKey, versionId, contentMD5 } = params;
+    return new ActionQueueEntry({
+        target: {
+            key: `test-bucket-name/${objectKey}`,
+            version: versionId,
+            contentMd5: contentMD5 || 'd41d8cd98f00b204e9800998ecf8427e',
+        }
+    });
+}
+
+function processingFunc(entry, cb) {
+    setTimeout(() => {
+        // eslint-disable-next-line no-param-reassign
+        entry.object.value = entry.setValueTo;
+        cb();
+    }, Math.random() * 100);
+}
+
+const replicationTaskScheduler = new ReplicationTaskScheduler(processingFunc);
+
+const queueEntryClasses = [
+    'ObjectQueueEntry',
+    'ActionQueueEntry',
+];
+
+queueEntryClasses.forEach(mockClass => {
+    let objects = [];
+    let doneCount = 0;
+    let queueEntryClass;
+
+    function getMockEntry(params) {
+        if (queueEntryClass === 'ObjectQueueEntry') {
+            return getMockQueueEntry(params);
+        }
+        if (queueEntryClass === 'ActionQueueEntry') {
+            return getMockActionEntry(params);
+        }
+        return undefined;
+    }
+
+    function scheduleDuplicateVersions(params, cb) {
+        const { uniqueKeyCount, duplicateKeyCount } = params;
+
+        for (let i = 0; i < uniqueKeyCount; ++i) {
+            objects.push({ objectKey: `key_${i}`, value: -1 });
+            // the following inner operations shall be executed in
+            // order because they have the same objectKey (passed
+            // to replicationTaskScheduler.push())
+            const versionId = uuid();
+            for (let j = 0; j < duplicateKeyCount; ++j) {
```

Comment:
minor: looks a little confusing to have `duplicateKeyCount` include the original key (so duplicate == 1 means no duplicate, right), maybe simply shift its meaning by one and update loops etc. accordingly.

Also to add more fuzziness in the test (in order to increase coverage of cases) you could pre-create an array of entries, shuffle it, then insert entries in the shuffled order (or pick entries at random and remove them from the array as it goes).

Reply:
Sounds good. Updated the `duplicateKeyCount` value. I'll follow with a way to randomize the entries.